server/controllers/userControllers.ts

```diff
@@ -1,31 +1,76 @@
 import { Request, Response } from "express";
 import { DatabaseUser } from "../services/database.user";
 import { UserRequest } from "../types/requests/userRequest";
-// import { Session } from "express-session";
+import { hashPassword, isPassword } from "../utils/hashing";
+import { ObjectId } from "mongoose";
+import { StatusCodes } from "http-status-codes";
+
 declare module "express-session" {
 	interface SessionData {
 		isLoggedIn?: boolean;
+		userid?: ObjectId;
 		// Add any other custom properties you need
 	}
 }
 
 export const login = async (req: Request, res: Response) => {
-	// TODO: implement login
-	res.send("inside login");
+	const { username, password } = req.body;
+	// check if username exists and retrieve the hashed password
+	const user = await DatabaseUser.readByUsername(username);
+	if (user === undefined) {
+		res
+			.status(StatusCodes.FORBIDDEN)
+			.send("username password combo is incorrect");
+		return;
+	}
+	const isAuthenticated: boolean | undefined = await isPassword(
+		password,
+		user.password
+	);
+	if (typeof isAuthenticated === "undefined") {
+		res
+			.status(StatusCodes.INTERNAL_SERVER_ERROR)
+			.send("error comparing passwords");
+		
+	} else {
+		if (!isAuthenticated) {
+			res
+				.status(StatusCodes.FORBIDDEN)
+				.send("username password combo is incorrect");
+			return;
+		}
+		// log the user in
+		req.session.isLoggedIn = true;
+		// set the user id in the session
+		req.session.userid = user.id;
+		res.status(StatusCodes.OK).send("successfully logged in");
+	}
 };
 
 export const register = async (req: Request, res: Response) => {
-	const username = req.body.username;
-	const password = req.body.password;
-	// const location = req.body?.location;
-	// TODO check user table so that user dne
-	const newuser: UserRequest = { username, password };
+	const { username, password } = req.body;
+	// const location = req.body?.location
+	// check if the username already exists
+	const isDuplicate = (await DatabaseUser.readByUsername(username)) !== undefined;
```

Comment:
That function already returns a `user` if they exists, otherwise it returns `undefined`. I personally think its a lot cleaner simply checking if the user is undefined. 
```
const user= await DatabaseUser.readByUsername(username);
if (!user) {
	res.status(StatusCodes.CONFLICT).send("username already exists");
	return;
}
```